### PR TITLE
[vulkan] Support for multiple SNode trees in Vulkan

### PR DESCRIPTION
**Pull request by AmesingFlank**

### CODE REVIEW

File: taichi/backends/vulkan/runtime.cpp [hunk 1]
```diff
@@ -442,26 +467,41 @@ class VkRuntime ::Impl {
...
+
     cmdlist->buffer_fill(global_tmps_buffer_->get_ptr(0), gtmp_buffer_size,
                          /*data=*/0);
     stream->submit_synced(cmdlist.get());
   }
 
-  const SNodeDescriptorsMap *const snode_descriptors_;
+  void add_root_buffer() {
+#pragma message("Vulkan buffers size hardcoded")
+    size_t root_buffer_size = 64 * 1024 * 1024;
```
Comment (bobcao3) [inline taichi/backends/vulkan/runtime.cpp:478]: After materializing the SNode tree, we should be able to get a size for that tree?

Comment (bobcao3) [reply taichi/backends/vulkan/runtime.cpp:478]: Since we are not handling sparse yet

**Review by bobcao3 — APPROVED** [review]
Super great work!

File: taichi/backends/vulkan/codegen_vulkan.cpp [hunk 2]
```diff
@@ -70,9 +72,18 @@ class TaskCodegen : public IRVisitor {
     allow_undefined_visitor = true;
     invoke_default_visitor = true;
 
+    fill_node_to_root();
     ir_ = std::make_shared<spirv::IRBuilder>(params.device);
   }
 
+  void fill_node_to_root() {
```
Comment (k-ye) [inline taichi/backends/vulkan/codegen_vulkan.cpp:79]: Nit: snode

File: taichi/backends/vulkan/codegen_vulkan.cpp [hunk 3]
```diff
@@ -239,20 +254,22 @@ class TaskCodegen : public IRVisitor {
   void visit(SNodeLookupStmt *stmt) override {
     // TODO: SNodeLookupStmt -> GetSNodeCellStmt ?
     bool is_root{false};  // Eliminate first root snode access
+    int root_id = node_to_root_.at(stmt->snode->id);
```
Comment (k-ye) [inline taichi/backends/vulkan/codegen_vulkan.cpp:257]: nit: const

File: taichi/backends/vulkan/codegen_vulkan.cpp [hunk 4]
```diff
@@ -1224,16 +1245,19 @@ class TaskCodegen : public IRVisitor {
...
 
-    auto bind_buffer = [&](TaskAttributes::Buffers buffer) {
+    auto bind_buffer = [&](BufferInfo buffer) {
       result.push_back({buffer, binding});
       buffer_binding_map_[buffer] = binding++;
     };
 
-    bind_buffer(BuffersEnum::Root);
-    bind_buffer(BuffersEnum::GlobalTmps);
+    for (int root = 0; root < compiled_structs_.size(); ++root) {
```
Comment (k-ye) [inline taichi/backends/vulkan/codegen_vulkan.cpp:1253]: BTW, if a kernel only uses the second root, will this still work? (I.e. do we still need to bind the first root?)

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/backends/vulkan/codegen_vulkan.cpp (hunk 4 above)
Comment (bobcao3) [reply taichi/backends/vulkan/codegen_vulkan.cpp:1253]: I think it's fine, as the descriptor layout is locked.

File: taichi/backends/vulkan/runtime.cpp (hunk 1 above)
Comment (AmesingFlank) [reply taichi/backends/vulkan/runtime.cpp:478]: hmm yeah. lemme try.

File: taichi/backends/vulkan/codegen_vulkan.cpp (hunk 4 above)
Comment (AmesingFlank) [reply taichi/backends/vulkan/codegen_vulkan.cpp:1253]: This will work, but there will indeed be an extra unused binding in that case.

**Pull request merged**